### ISSUE TITLE
allow use of `typeof this.xxx` (Error: Cannot find name this)

### ISSUE THREAD

**Issue by xahon** [palantir/tslint#4793]
### Bug Report

- __TSLint version__:     "tslint": "^5.18.0",
- __TypeScript version__:     "typescript": "3.5.3"
- __Running TSLint via__: Rider/WebStorm

#### TypeScript code being linted

```ts
    // it works
    const { state } = (this)
    type inputsType = typeof state.form.inputs & { [s: string]: string }

    // it doesn't
    type inputsType = typeof this.state.form.inputs & { [s: string]: string }
                               ^ error here
```
![image](https://user-images.githubusercontent.com/18731681/61054857-aaf54980-a433-11e9-95fc-b77ce165baad.png)

I want do it for better typing when merging reducer states
```ts
{
    ...this.state,
    form: {
       ...this.state.form,
       inputs: Object
         .keys(this.state.form)
         .reduce((obj, key) => {
              obj[key] = ""
              return obj
           },
           { ...this.state.form.inputs } as any as inputsType, // here
         ),
    },
},
```

#### Actual behavior
error

#### Expected behavior
no error

**Comment by JoshuaKGoldberg** [comment]
@xahon it looks like your issue is with **TypeScript**, not TSLint. TypeScript is the programming language and compiler that's giving you a type error in your code. TSLint is a set of analysis that uses TypeScript.
You'll want to look for this issue on https://github.com/Microsoft/TypeScript/issues. Enjoy!

**Comment by xahon** [comment]
Ok, sorry, there is already the issue https://github.com/microsoft/TypeScript/issues/1554
